### PR TITLE
TensorFlow/Keras/TensorBoard: add v2.14.0

### PR DESCRIPTION
**Pull request by adamjstewart**
* https://github.com/tensorflow/tensorflow/releases/tag/v2.14.0
* https://github.com/keras-team/keras/releases/tag/v2.14.0
* https://github.com/tensorflow/tensorboard/releases/tag/2.14.0

### CODE REVIEW

**Comment by adamjstewart** [comment]
@spackbot run pipeline

**Review by tldahlgren — DISMISSED** [review]
Confirmed the homepage (of the new package) and version sha256s.

**Comment by adamjstewart** [comment]
@spackbot fix style

File: var/spack/repos/builtin/packages/py-ml-dtypes/package.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+from spack.package import *
+
+
+class PyMlDtypes(PythonPackage):
+    """A stand-alone implementation of several NumPy dtype extensions
+    used in machine learning libraries."""
+
+    homepage = "https://github.com/jax-ml/ml_dtypes"
+    pypi = "ml_dtypes/ml_dtypes-0.3.1.tar.gz"
+
```
Comment (aweits) [inline var/spack/repos/builtin/packages/py-ml-dtypes/package.py:17]:
```diff
diff --git a/var/spack/repos/builtin/packages/py-ml-dtypes/package.py b/var/spack/repos/builtin/packages/py-ml-dtypes/package.py
index 6527c34ba99..a4dc53ff18e 100644
--- a/var/spack/repos/builtin/packages/py-ml-dtypes/package.py
+++ b/var/spack/repos/builtin/packages/py-ml-dtypes/package.py
@@ -11,10 +11,11 @@ class PyMlDtypes(PythonPackage):
     used in machine learning libraries."""
 
     homepage = "https://github.com/jax-ml/ml_dtypes"
+    git = "https://github.com/jax-ml/ml_dtypes"
     pypi = "ml_dtypes/ml_dtypes-0.3.1.tar.gz"
 
-    version("0.3.1", sha256="60778f99194b4c4f36ba42da200b35ef851ce4d4af698aaf70f5b91fe70fc611")
-    version("0.2.0", sha256="6488eb642acaaf08d8020f6de0a38acee7ac324c1e6e92ee0c0fea42422cb797")
+    version("0.3.1", tag="v0.3.1", submodules=True)
+    version("0.2.0", tag="v0.2.0", submodules=True)
 
     depends_on("python@3.9:", when="@0.3:", type=("build", "link", "run"))
     depends_on("py-numpy@1.21:", type=("build", "link", "run"))
```

Comment (aweits) [reply var/spack/repos/builtin/packages/py-ml-dtypes/package.py:17]: avoids this error due to files not present in tarball
```
  In file included from third_party/eigen/Eigen/Core:205,
                   from ml_dtypes/_src/dtypes.cc:32:
  third_party/eigen/Eigen/src/Core/arch/AVX512/TrsmKernel.h:51:10: fatal error: TrsmUnrolls.inc: No such file or directory
     51 | #include "TrsmUnrolls.inc"
        |          ^~~~~~~~~~~~~~~~~
```

**Comment by aweits** [comment]
This seems to get the keras build unbroken - there might be other hidden fallout from hermetic python though, needs more examination...

```diff
diff --git a/var/spack/repos/builtin/packages/py-tensorflow/package.py b/var/spack/repos/builtin/packages/py-tensorflow/package.py
index eddbfa919ad..d6f9ef84128 100644
--- a/var/spack/repos/builtin/packages/py-tensorflow/package.py
+++ b/var/spack/repos/builtin/packages/py-tensorflow/package.py
@@ -516,6 +516,8 @@ def setup_build_environment(self, env):
         # Please input the desired Python library path to use
         env.set("PYTHON_LIB_PATH", python_platlib)
 
+        env.set("TF_PYTHON_VERSION", str(spec["python"].version.up_to(2)))
+
         # Ensure swig is in PATH or set SWIG_PATH
         env.set("SWIG_PATH", spec["swig"].prefix.bin.swig)
 
```

**Review by alalazo — APPROVED** [review]

**Pull request merged**